src/main/java/net/neoforged/neoforge/transfer/energy/ItemAccessEnergyHandler.java

```diff
@@ -0,0 +1,168 @@
+        this(itemAccess, energyComponent, capacity, maxTransfer, maxTransfer);
+    }
+
+    /**
+     * Creates a new {@link ItemAccessEnergyHandler} instance.
+     *
+     * @param itemAccess      item access backing this handler
+     * @param energyComponent type of the data component used to store the energy amount
+     * @param capacity        maximum capacity of the handler
+     * @param maxInsert       maximum amount that can be accepted in each call to {@link #insert}
+     * @param maxExtract      maximum amount that can be accepted in each call to {@link #extract}
+     * @throws IllegalArgumentException if any of the int arguments is negative
+     */
+    public ItemAccessEnergyHandler(ItemAccess itemAccess, DataComponentType<Integer> energyComponent, int capacity, int maxInsert, int maxExtract) {
+        TransferPreconditions.checkNonNegative(capacity);
+        TransferPreconditions.checkNonNegative(maxInsert);
+        TransferPreconditions.checkNonNegative(maxExtract);
+
+        this.itemAccess = itemAccess;
+        // Store the current item, such that if the item changes later we don't return any stored content from it.
+        this.validItem = itemAccess.getResource().getItem();
+        this.energyComponent = energyComponent;
+        this.capacity = capacity;
+        this.maxInsert = maxInsert;
+        this.maxExtract = maxExtract;
+    }
+
+    /**
+     * Retrieves the amount stored in the {@linkplain ItemAccess#getResource() current contents} of the item access.
+     */
+    protected int getAmountFrom(ItemResource accessResource) {
+        return accessResource.getOrDefault(energyComponent, 0);
+    }
+
+    /**
+     * Returns a resource with updated amount.
+     *
+     * @param accessResource current resource, before the update
+     * @param newAmount      the new amount
+     * @return {@code accessResource} updated with the new amount,
+     *         or {@link ItemResource#EMPTY} if the new amount cannot be stored
+     * @implNote This function <strong>should not</strong> mutate the {@linkplain #itemAccess item access},
+     *           that will be done by the calling code based on the results of this function.
+     */
+    protected ItemResource update(ItemResource accessResource, int newAmount) {
+        return accessResource.with(energyComponent, newAmount);
+    }
+
+    @Override
+    public long getAmountAsLong() {
+        var accessResource = itemAccess.getResource();
+        if (!accessResource.is(validItem)) {
+            return 0;
+        }
+
+        // Cast as the product of two ints might overflow an int, but fits into a long
+        return (long) itemAccess.getAmount() * getAmountFrom(itemAccess.getResource());
+    }
+
+    @Override
+    public long getCapacityAsLong() {
+        var accessResource = itemAccess.getResource();
+        if (!accessResource.is(validItem)) {
```

Comment:
Please extract this check for if the backing access is still valid into a private/protected helper method so that it is more clear for here and the get amount that it is validating the backing access is valid. Also do we need to be checking it for validity inside our `insert` and `extract` impls?